### PR TITLE
`Development`: Remove use of deprecated methods

### PR DESCRIPTION
**Pull request by krusche**
This PR replaces deprecated methods (in particular ResponseUtil) on the server with exceptions. Please focus on reviewing the code!

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.

#### Code Review
- [x] Review 1
- [x] Review 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java [hunk 1]
```diff
@@ -694,21 +690,15 @@ private void prepareComplaintAndSubmission(Complaint complaint, Submission submi
...
 
         List<Submission> submissions = new ArrayList<>();
-
         for (StudentParticipation participation : studentParticipationPage.getContent()) {
             Optional<Submission> optionalSubmission = participation.findLatestSubmission();
-
-            if (!optionalSubmission.isEmpty()) {
-                submissions.add(optionalSubmission.get());
-            }
+            optionalSubmission.ifPresent(submissions::add);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java:698]: I think we can replace this whole loop:
```
List<Submission> submissions = studentParticipationPage.getContent().stream()
    .map(Participation::findLatestSubmission())
    .filter(Optional::isPresent)
    .toList();
```

**Review by Strohgelaender — DISMISSED** [review]
I went through all files and did not spot any mistakes, the changes look good. I only added a small suggestion:

**Review by coaxial1 — APPROVED** [review]
Besides the comment above (which I second), I did not spot any mistakes. LGTM.

**Review by b-fein — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamAccessService.java [hunk 2]
```diff
@@ -99,47 +85,39 @@ public StudentExamAccessService(CourseRepository courseRepository, UserRepositor
...
      * @param examId        the id of the exam
      * @param currentUser   the user
-     * @param isTestRun       flag to determine if this is a testRun
-     * @param <T>           The type of the return type of the requesting route so that the
-     *                      response can be returned there
-     * @return an Optional with a typed ResponseEntity. If it is empty all checks passed
+     * @param isTestRun     flag to determine if this is a testRun
      */
-    public <T> Optional<ResponseEntity<T>> checkCourseAndExamAccess(Long courseId, Long examId, User currentUser, boolean isTestRun) {
+    public void checkCourseAndExamAccess(Long courseId, Long examId, User currentUser, boolean isTestRun) {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamAccessService.java:90]:
```suggestion
    public void checkCourseAndExamAccessElseThrow(Long courseId, Long examId, User currentUser, boolean isTestRun) {
```

File: src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamAccessService.java [hunk 3]
```diff
@@ -64,33 +61,22 @@ public StudentExamAccessService(CourseRepository courseRepository, UserRepositor
...
-     * @param <T>           The type of the return type of the requesting route so that the
-     *                      response can be returned there
-     * @return an Optional with a typed ResponseEntity. If it is empty all checks passed
      */
-    public <T> Optional<ResponseEntity<T>> checkStudentExamAccess(Long courseId, Long examId, Long studentExamId, User currentUser, boolean isTestRun) {
-        Optional<ResponseEntity<T>> courseAndExamAccessFailure = checkCourseAndExamAccess(courseId, examId, currentUser, isTestRun);
-        if (courseAndExamAccessFailure.isPresent()) {
-            return courseAndExamAccessFailure;
-        }
+    public void checkStudentExamAccess(Long courseId, Long examId, Long studentExamId, User currentUser, boolean isTestRun) {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamAccessService.java:65]:
```suggestion
    public void checkStudentExamAccessElseThrow(Long courseId, Long examId, Long studentExamId, User currentUser, boolean isTestRun) {
```

File: src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamAccessService.java [hunk 4]
```diff
@@ -47,13 +48,9 @@ public StudentExamAccessService(CourseRepository courseRepository, UserRepositor
...
      * @param studentExamId the id of the student exam
      * @param isTestRun     flag to determine if it is a test run or not
-     * @param <T>           The type of the return type of the requesting route so that the
-     *                      response can be returned there
-     * @return an Optional with a typed ResponseEntity. If it is empty all checks passed
      */
-    public <T> Optional<ResponseEntity<T>> checkStudentExamAccess(Long courseId, Long examId, Long studentExamId, boolean isTestRun) {
-        User currentUser = userRepository.getUserWithGroupsAndAuthorities();
-        return checkStudentExamAccess(courseId, examId, studentExamId, currentUser, isTestRun);
+    public void checkStudentExamAccess(Long courseId, Long examId, Long studentExamId, boolean isTestRun) {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamAccessService.java:52]:
```suggestion
    public void checkStudentExamAccessElseThrow(Long courseId, Long examId, Long studentExamId, boolean isTestRun) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java [hunk 5]
```diff
@@ -205,18 +206,17 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
...
                 return ResponseEntity.ok().build();
             }
         }
-        if (lastGraded && submission.get().getType() != SubmissionType.INSTRUCTOR && submission.get().getType() != SubmissionType.TEST
-                && exerciseDateService.isAfterDueDate(participation)) {
+        if (lastGraded && submission.getType() != SubmissionType.INSTRUCTOR && submission.getType() != SubmissionType.TEST && exerciseDateService.isAfterDueDate(participation)) {
             // If the submission is not the latest but the last graded, there is no point in triggering the build again as this would build the most recent VCS commit.
             // This applies only to students submissions after the exercise due date.
-            return notFound();
+            throw new EntityNotFoundException("Cannot trigger failed build. There is a submission after exercise duet date");
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:216]:
```suggestion
            throw new EntityNotFoundException("Cannot trigger failed build. There is a submission after the exercise due date");
```

**Review by b-fein — CHANGES_REQUESTED** [review]
I think there are a few more methods which should get the `ElseThrow` suffix. And I found a typo. Looks good otherwise.

**Review by b-fein — APPROVED** [review]
The changes since the last review look good. I did not see anything else during the re-review. :+1:

**Pull request merged**